File: android/modules/pokemon-sdk/src/main/java/com/multiplatformbusinesssdk/pokemon/PokemonSDK.kt
package com.multiplatformbusinesssdk.pokemon

import com.multiplatformbusinesssdk.core.ApiRequest
import com.multiplatformbusinesssdk.core.HttpMethod
import com.multiplatformbusinesssdk.core.NetworkEngine

object PokemonSDKConfig {
    const val baseUrl: String = "https://pokeapi.co/api/v2"
}

data class PokemonListResponse(
    val count: Int,
    val results: List<PokemonEntry>
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): PokemonListResponse {
                        val count = json.getInt("count")
            val resultsArray = json.getJSONArray("results")
                        val results = buildList {
                            for (index in 0 until resultsArray.length()) {
                                add(PokemonEntry.fromJson(resultsArray.getJSONObject(index)))
                            }
                        }
            return PokemonListResponse(count, results)
        }
    }
}

data class PokemonEntry(
    val name: String,
    val url: String
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): PokemonEntry {
                        val name = json.getString("name")
            val url = json.getString("url")
            return PokemonEntry(name, url)
        }
    }
}

data class PokemonDetailResponse(
    val id: Int,
    val name: String,
    val height: Int,
    val weight: Int
) {
    companion object {
        fun fromJson(json: org.json.JSONObject): PokemonDetailResponse {
                        val id = json.getInt("id")
            val name = json.getString("name")
            val height = json.getInt("height")
            val weight = json.getInt("weight")
            return PokemonDetailResponse(id, name, height, weight)
        }
    }
}


private class FetchPokemonListRequest(private val limit: Int = 20, private val offset: Int = 0) : ApiRequest<PokemonListResponse> {
    override val path: String = "/pokemon"
    override val method: HttpMethod = HttpMethod.GET
    override val body: ByteArray? = null
    override val headers: Map<String, String> = mapOf("Accept" to "application/json")
    override val query: Map<String, String> = mapOf(
            "limit" to limit.toString(),
            "offset" to offset.toString()
        )
    override val requiresAuthorization: Boolean = false

    override fun parse(payload: String): PokemonListResponse =
        PokemonListResponse.fromJson(org.json.JSONObject(payload))
}

private class FetchPokemonDetailRequest(private val name: String) : ApiRequest<PokemonDetailResponse> {
    override val path: String = "/pokemon/$name"
    override val method: HttpMethod = HttpMethod.GET
    override val body: ByteArray? = null
    override val headers: Map<String, String> = mapOf("Accept" to "application/json")
    override val query: Map<String, String> = emptyMap()
    override val requiresAuthorization: Boolean = false

    override fun parse(payload: String): PokemonDetailResponse =
        PokemonDetailResponse.fromJson(org.json.JSONObject(payload))
}


interface PokemonRepository {
    suspend fun fetchPokemonList(limit: Int = 20, offset: Int = 0): PokemonListResponse
    suspend fun fetchPokemonDetail(name: String): PokemonDetailResponse
}

class DefaultPokemonRepository(
    private val networkEngine: NetworkEngine
) : PokemonRepository {
override suspend fun fetchPokemonList(limit: Int = 20, offset: Int = 0): PokemonListResponse {
    return networkEngine.request(FetchPokemonListRequest(limit = limit, offset = offset))
}

override suspend fun fetchPokemonDetail(name: String): PokemonDetailResponse {
    return networkEngine.request(FetchPokemonDetailRequest(name = name))
}

}

File: BusinessSDK/Sources/PokemonSDK/PokemonSDK.swift
import Foundation
import BusinessSDKCore

public enum PokemonSDKConfig {
    public static let baseURL = URL(string: "https://pokeapi.co/api/v2")!
}

public struct PokemonListResponse: Codable, Sendable {
    public let count: Int
    public let results: [PokemonEntry]

    public init(count: Int, results: [PokemonEntry]) {
        self.count = count
        self.results = results
    }
}

public struct PokemonEntry: Codable, Sendable {
    public let name: String
    public let url: String

    public init(name: String, url: String) {
        self.name = name
        self.url = url
    }
}

public struct PokemonDetailResponse: Codable, Sendable {
    public let id: Int
    public let name: String
    public let height: Int
    public let weight: Int

    public init(id: Int, name: String, height: Int, weight: Int) {
        self.id = id
        self.name = name
        self.height = height
        self.weight = weight
    }
}


private struct PokemonRepositoryFetchPokemonListRequest: APIRequest {
    private let limit: Int
    private let offset: Int

    init(limit: Int = 20, offset: Int = 0) {
        self.limit = limit
        self.offset = offset
    }

    let method: HTTPMethod = .get
    let body: Data? = nil
    let headers: [String: String] = ["Accept": "application/json"]
    let requiresAuthorization: Bool = false

    var path: String {
        "/pokemon"
    }

    var queryItems: [URLQueryItem] {
        [URLQueryItem(name: "limit", value: String(describing: limit)), URLQueryItem(name: "offset", value: String(describing: offset))]
    }

    typealias Response = PokemonListResponse
}

private struct PokemonRepositoryFetchPokemonDetailRequest: APIRequest {
    private let name: String

    init(name: String) {
        self.name = name
    }

    let method: HTTPMethod = .get
    let body: Data? = nil
    let headers: [String: String] = ["Accept": "application/json"]
    let requiresAuthorization: Bool = false

    var path: String {
        "/pokemon/\(name)"
    }

    var queryItems: [URLQueryItem] {
        []
    }

    typealias Response = PokemonDetailResponse
}


public protocol PokemonRepositoryProtocol: Sendable {
    func fetchPokemonList(limit: Int, offset: Int) async throws -> PokemonListResponse
    func fetchPokemonDetail(name: String) async throws -> PokemonDetailResponse
}

public final class PokemonRepository: PokemonRepositoryProtocol, @unchecked Sendable {
    private let networkEngine: NetworkEngineProtocol

    public init(networkEngine: NetworkEngineProtocol) {
        self.networkEngine = networkEngine
    }

public func fetchPokemonList(limit: Int = 20, offset: Int = 0) async throws -> PokemonListResponse {
    try await networkEngine.request(PokemonRepositoryFetchPokemonListRequest(limit: limit, offset: offset))
}

public func fetchPokemonDetail(name: String) async throws -> PokemonDetailResponse {
    try await networkEngine.request(PokemonRepositoryFetchPokemonDetailRequest(name: name))
}

}


public protocol PokemonSDKClientProtocol: Sendable {
    var pokemonRepository: PokemonRepositoryProtocol { get }
}

public struct PokemonSDKClient: PokemonSDKClientProtocol, Sendable {
    public let pokemonRepository: PokemonRepositoryProtocol

    public init(networkEngine: NetworkEngineProtocol = NetworkEngine(baseURL: PokemonSDKConfig.baseURL)) {
        self.pokemonRepository = PokemonRepository(networkEngine: networkEngine)
    }
}
